temporalio/lib/temporalio/internal/worker/activity_worker.rb

```diff
@@ -0,0 +1,345 @@
+          @activities = worker.options.activities.each_with_object({}) do |act, hash|
+            # Class means create each time, instance means just call, definition
+            # does nothing special
+            defn = Activity::Definition.from_activity(act)
+            # Confirm name not in use
+            raise ArgumentError, "Multiple activities named #{defn.name}" if hash.key?(defn.name)
+
+            # Confirm executor is a known executor and let it initialize
+            executor = worker.options.activity_executors[defn.executor]
+            raise ArgumentError, "Unknown executor '#{defn.executor}'" if executor.nil?
+
+            executor.initialize_activity(defn)
+
+            hash[defn.name] = defn
+          end
+
+          # Need mutex for the rest of these
+          @running_activities_mutex = Mutex.new
+          @running_activities = {}
+          @running_activities_empty_condvar = ConditionVariable.new
+        end
+
+        def set_running_activity(task_token, activity)
+          @running_activities_mutex.synchronize do
+            @running_activities[task_token] = activity
+          end
+        end
+
+        def get_running_activity(task_token)
+          @running_activities_mutex.synchronize do
+            @running_activities[task_token]
+          end
+        end
+
+        def remove_running_activity(task_token)
+          @running_activities_mutex.synchronize do
+            @running_activities.delete(task_token)
+            @running_activities_empty_condvar.broadcast if @running_activities.empty?
+          end
+        end
+
+        def wait_all_complete
+          @running_activities_mutex.synchronize do
+            @running_activities_empty_condvar.wait(@running_activities_mutex) until @running_activities.empty?
+          end
+        end
+
+        def handle_task(task)
+          @scoped_logger.debug("Received activity task: #{task}") if LOG_TASKS
+          if !task.start.nil?
+            handle_start_task(task.task_token, task.start)
+          elsif !task.cancel.nil?
+            handle_cancel_task(task.task_token, task.cancel)
+          else
+            raise "Unrecognized activity task: #{task}"
+          end
+        end
+
+        def handle_start_task(task_token, start)
+          set_running_activity(task_token, nil)
+
+          # Find activity definition
+          defn = @activities[start.activity_type]
+          if defn.nil?
+            raise Error::ApplicationError.new(
+              "Activity #{start.activity_type} for workflow #{start.workflow_execution.workflow_id} " \
+              "is not registered on this worker, available activities: #{@activities.keys.sort.join(', ')}",
+              type: 'NotFoundError'
+            )
+          end
+
+          # Run everything else in the excecutor
+          executor = @worker.options.activity_executors[defn.executor]
+          executor.execute_activity(defn) do
```

Comment:
nit: Looks like `executor.execute_activity` has no real need for `defn`

Reply:
Only because you're looking at the two implementations today :-) This is a user-facing abstraction and I am concerned about a future where people want to configure their executor on a per-activity basis (granted they may have to extend activity definition or something). I considered not even exposing this abstraction, but there are enough users of different async libraries out there that it seemed worth it. And it seemed wrong to tell an executor to execute a block for an activity without them even being able to know what activity it is, especially if I have it right here. Similar to how our recent worker slot supplier abstraction provides info to make decisions there even though we don't use all of that info in our limited implementations today.